clients/src/main/java/org/apache/kafka/common/utils/ChildFirstClassLoader.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.utils;
+
+import org.apache.kafka.common.KafkaException;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.Locale;
+import java.util.NoSuchElementException;
+
+/**
+ * A class loader that looks for classes and resources in a specified class path first, before delegating to its parent
+ * class loader.
+ */
+public class ChildFirstClassLoader extends URLClassLoader {
```

Comment:
For other broker side plugins (e.g. Authorizer), we never needed a special class loader. Why do we need this for the remote storage plugin?

Reply:
This is to avoid library conflicts directly or indirectly with the existing libraries in the system classpath. This classloader loads the libraries in the given paths. Whenever a class needs to be loaded, it delegates to the system classloader only when it is not found in the configured paths. RSM or RLMM providers can have all the required libraries in given directories and add them as classpath for RSM or RLMM. This provides isolation with the libraries in the system classpath.

Reply:
Yes, I understand the intent. However, there are quite a few pluggable components on the broker side right now. Sorting out the classpath dependencies among all of them seems quite complicated. Another way to solve the issue is through shading. A potential conflicting library could be renamed to a different package.

Reply:
+1 @junrao